File: solutions/lua/triangle/1/triangle.lua
local triangle = {}

function triangle.kind(a, b, c)
    if (a <= 0 or b <= 0 or c <= 0) then
      error('Input Error')
    end

    if (not ((a+b) >= c and (b+c) >= a and (a+c) >= b)) then
      error('Input Error')
    end
    
    if (a == b and b == c) then
      return 'equilateral'
    end

    if ( a == b or b == c or c == a) then
      return 'isosceles'
    end
    
    return 'scalene'
end

return triangle

File: solutions/ruby/triangle/1/triangle.rb
=begin
Write your code for the 'Triangle' exercise in this file. Make the tests in
`triangle_test.rb` pass.

To get started with TDD, see the `README.md` file in your
`ruby/triangle` directory.
=end
class Triangle
  def initialize(args)
    @first, @second, @third = args
  end

  def equilateral?
    return false unless is_valid?
    @first == @second || @second == @third
  end

  def scalene?
    return false unless is_valid?
    @first != @second && @second != @third && @third != @first
  end

  def isosceles?
    return false unless is_valid?
    @first == @second || @second == @third || @third == @first
  end

  def any_zero?
    @first.zero? || @second.zero? || @third.zero?
  end

  def is_valid?
    return false if any_zero?

    @first + @second >= @third &&
    @second + @third >= @first &&
    @first + @third >= @second
  end
end